src/trainer_hill.c

```diff
@@ -0,0 +1,1121 @@
+    FreeDataStruct();
+}
+
+bool32 InTrainerHill(void)
+{
+    bool32 ret;
+
+    if (gMapHeader.mapLayoutId == 0x19F
+            || gMapHeader.mapLayoutId == 0x1A0
+            || gMapHeader.mapLayoutId == 0x1A1
+            || gMapHeader.mapLayoutId == 0x1A2)
+        ret = TRUE;
+    else
+        ret = FALSE;
+
+    return ret;
+}
+
+u8 GetCurrentTrainerHillMapId(void)
+{
+    u8 ret;
+
+    if (gMapHeader.mapLayoutId == 0x19F)
+        ret = 1;
+    else if (gMapHeader.mapLayoutId == 0x1A0)
+        ret = 2;
+    else if (gMapHeader.mapLayoutId == 0x1A1)
+        ret = 3;
+    else if (gMapHeader.mapLayoutId == 0x1A2)
+        ret = 4;
+    else if (gMapHeader.mapLayoutId == 0x1A3)
+        ret = 5;
+    else if (gMapHeader.mapLayoutId == 0x19E)
+        ret = 6;
+    else
+        ret = 0;
+
+    return ret;
+}
+
+static bool32 sub_81D6100(void)
+{
+    bool32 ret;
+
+    if (gMapHeader.mapLayoutId == 0x1A3)
+        ret = TRUE;
+    else
+        ret = FALSE;
+
+    return ret;
+}
+
+const struct WarpEvent* sub_81D6120(void)
+{
+    const struct MapHeader *header = Overworld_GetMapHeaderByGroupAndId(MAP_GROUP(TRAINER_HILL_4F), MAP_NUM(TRAINER_HILL_4F));
+
+    return &header->events->warps[1];
+}
+
+const struct WarpEvent* sub_81D6134(u8 warpEventId)
+{
+    u8 id;
+    const struct MapHeader *header;
+
+    if (warpEventId == 1)
+        return &gMapHeader.events->warps[1];
+
+    id = sub_81D6490();
+    if ((u8)(id - 1) >= 4)
+        id = 4;
+
+    header = Overworld_GetMapHeaderByGroupAndId(MAP_GROUP(TRAINER_HILL_4F), gUnknown_0862A688[id - 1]);
+    return &header->events->warps[0];
+}
+
+u16 LocalIdToHillTrainerId(u8 localId)
+{
+    return gSaveBlock2Ptr->frontier.field_CB4[localId - 1];
+}
+
+bool8 GetHillTrainerFlag(u8 eventObjectId)
+{
+    u32 floorId = GetFloorId() * 2;
+    u8 bitId = gEventObjects[eventObjectId].localId - 1 + floorId;
+
+    return gSaveBlock2Ptr->frontier.field_EE0 & gBitTable[bitId];
+}
+
+void SetHillTrainerFlag(void)
+{
+    u8 i;
+    u8 floorId = GetFloorId() * 2;
+
+    for (i = 0; i < 2; i++)
+    {
+        if (gSaveBlock2Ptr->frontier.field_CB4[i] == gTrainerBattleOpponent_A)
+        {
+            gSaveBlock2Ptr->frontier.field_EE0 |= gBitTable[floorId + i];
+            break;
+        }
+    }
+
+    if (gBattleTypeFlags & BATTLE_TYPE_TWO_OPPONENTS)
+    {
+        for (i = 0; i < 2; i++)
+        {
+            if (gSaveBlock2Ptr->frontier.field_CB4[i] == gTrainerBattleOpponent_B)
+            {
+                gSaveBlock2Ptr->frontier.field_EE0 |= gBitTable[floorId + i];
+                break;
+            }
+        }
+    }
+}
+
+const u8 *sub_81D62AC(void)
+{
+    return EventScript_2C83F0;
+}
+
+static void sub_81D62B4(void)
+{
+    CopyTrainerHillTrainerText(5, gSpecialVar_LastTalked);
+    sub_80982B8();
+}
+
+static void sub_81D62CC(u16 trainerId, u8 firstMonId)
+{
+    u8 trId, level;
+    s32 i, floorId, arrId;
+
+    if ((u16)(trainerId - 1) > 1)
```

Comment:
Probably `trainerId == 0 || trainerId > 2`, or something similar (these casts and subtractions in a comparison usually are)

Reply:
Good call.